hs/app/reachpc/ReachPC/Config.hs

```diff
@@ -42,113 +44,207 @@ readEnum :: forall a. (Show a, Read a, Enum a, Bounded a) => String -> String ->
+  -- project = "..." in reach.toml
+  let cfg_project = unwrapConfigOption "Unspecified project name." rtml_project
 
   return Config{..}
 
-readOrCreateGlobalReachToml :: IO RchToml
-readOrCreateGlobalReachToml = do
-  configTomlPath <- inConfigDir "reach.toml"
-  configTomlMissing <- not <$> doesFileExist configTomlPath
-  when configTomlMissing $ do
-    putStrLn $ "Generated global reach.toml at \"" <> configTomlPath <> "\""
-    putStrLn "You may want to run `reach config` or edit it yourself."
-    BS.writeFile configTomlPath defaultReachToml
-  inConfigDir "reach.toml" >>= readReachToml
-
-readOrErrorProjectReachToml :: IO RchToml
-readOrErrorProjectReachToml = fromMaybe (error "No reach.toml for a project was found.") <$> discoverProjectReachToml
-
-discoverProjectReachToml :: IO (Maybe RchToml)
-discoverProjectReachToml = getCurrentDirectory >>= go . tomlPaths
+readOrErrorReachToml :: IO (FilePath, ReachToml)
+readOrErrorReachToml = reachTomlPath >>= maybe notFound (\p -> sequence (p, T.decodeFile reachTomlCodec p))
+  where
+    notFound = error (error "No reach.toml was found.")
+
+readOrEmptyGlobalsToml :: IO GlobalsToml
+readOrEmptyGlobalsToml = do
+  path <- globalsTomlPath
+  missing <- not <$> doesFileExist path
+  tml <- if missing then return (GlobalsToml Nothing Nothing) else T.decodeFile globalsTomlCodec path
+  return tml
+
+-- Up-scan for a file called reach.toml (check ./reach.toml, ../reach.toml, ../../reach.toml etc)
+reachTomlPath :: IO (Maybe FilePath)
+reachTomlPath = go . tomlPaths =<< getCurrentDirectory
  where
   tomlPaths = map (</> "reach.toml") . reverse . scanl1 (</>) . splitPath
   splitPath = split . keepDelimsR . onSublist $ "/"
   go [] = return Nothing
   go (path:paths) = do
     exists <- doesFileExist path
-    if exists then Just <$> readReachToml path else go paths
-
-inConfigDir :: FilePath -> IO FilePath
-inConfigDir path = (</> path) <$> getXdgDirectory XdgConfig "reach"
-
-readReachToml :: FilePath -> IO RchToml
-readReachToml = T.decodeFile reachTomlCodec
-
--- TODO: currently this will translate a bad setting (eg cloud-or-local = "nonsense") into Nothing
--- instead of throwing an error and printing valid options. This can probably be fixed with T.match
-reachTomlCodec :: T.TomlCodec RchToml
-reachTomlCodec = RchToml
-  <$> T.dioptional (T.enumBounded "cloud-or-local") T..= rtml_cloudOrLocal
-  <*> T.dioptional (T.enumBounded "connector") T..= rtml_connector
-  <*> T.dioptional (T.arrayOf T._String "forward-env-vars") T..= rtml_forwardEnvVars
-  <*> T.dioptional (T.string "project") T..= rtml_project
-
--- TODO: put into separate file and include at compile time
-defaultReachToml :: ByteString
-defaultReachToml = $(makeRelativeToProject "app/reachpc/ReachPC/default_reach.toml" >>= embedFile)
+    if exists then return $ Just path else go paths
+
+globalsTomlPath :: IO FilePath
+globalsTomlPath = getXdgDirectory XdgConfig "reach" <&> (</> "global.toml")
```

Comment:
I think it should be `config.toml` not "global"... it's global because of where it is, not its name, right?

Reply:
true